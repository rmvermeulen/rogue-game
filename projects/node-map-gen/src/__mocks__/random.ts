import * as Chance from 'chance';

// use deterministic seed
export const random = new Chance(12345);
