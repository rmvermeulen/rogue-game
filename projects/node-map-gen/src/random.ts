import * as Chance from 'chance';

export const random = new Chance();
