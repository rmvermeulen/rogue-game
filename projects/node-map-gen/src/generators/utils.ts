import {
  anyPass,
  compose,
  CurriedFunction2,
  curry,
  filter,
  identity,
  map,
  prop,
  reject,
  uniqBy,
  unnest,
  whereEq,
} from 'ramda';
import { ICell } from '../grid';
import { random } from '../random';
import { generate } from '../utils';

export const withoutBy = curry(
  <T>(isEqual: (a: T, b: T) => boolean, banList: T[], list: T[]): T[] => {
    const isBanned = anyPass(map((a: T) => (b: T) => isEqual(a, b), banList));

    return reject(isBanned, list);
  },
);

type HasId = Readonly<{ id: number }>;

export const withoutById = withoutBy((a: HasId, b: HasId) => a.id === b.id);

export const gridFactory = <T>(
  width: number,
  height: number,
  create: (x: number, y: number, index: number) => T,
): T[] => {
  let index = 0;

  return unnest(
    generate(height, y =>
      generate(width, x => {
        const cell = create(x, y, index);
        index += 1;

        return cell;
      }),
    ),
  );
};

export type ProtoCell = Readonly<ICell & { neighborCount: number }>;

export type Point = Readonly<{ x: number; y: number }>;

export const isNeighborOf = (a: Point) => (b: Point): boolean =>
  Math.abs(a.x - b.x) + Math.abs(a.y - b.y) === 1;

export const findNeighborsFrom = <T extends Point>(sourceCells: T[]) => (
  currentCell: T,
): T[] => sourceCells.filter(isNeighborOf(currentCell));

export type Candidate = Point & Readonly<Record<'id' | 'room', number>>;

export const findCandidatesFrom = (sourceCells: Candidate[]) => (
  currentCells: Candidate[],
  excludeCurrent = false,
): Candidate[] => {
  return compose(
    excludeCurrent
      ? (withoutById(currentCells) as (cs: Candidate[]) => Candidate[])
      : identity,
    uniqBy(prop('id')),
    filter<Candidate>(whereEq({ room: -1 })),
    unnest,
    // chain(findNeighborsFrom(sourceCells)),
    map(findNeighborsFrom(sourceCells)),
  )(currentCells);
};

export const pickInitialRoomCells = (cells: ProtoCell[], roomCount: number) => {
  // pick a starting point for each room at random
  // - no overlap
  // - not locked in
  let pool = cells;
  const roomStartCells: ProtoCell[] = [];
  while (roomStartCells.length < roomCount) {
    const cell = random.pickone(pool);
    pool = withoutById([cell], pool);

    // cells outside the grid don't matter
    const neighborhood = cells.filter(isNeighborOf(cell));

    const pickedNeighbors = neighborhood.reduce(
      (total, neighbor) => total + (roomStartCells.includes(neighbor) ? 1 : 0),
      0,
    );

    if (neighborhood.length === pickedNeighbors) {
      // cell is locked in, all it's neighbors
      // are a starting point, or missing (outside grid)
      continue;
    }

    const roomId = roomStartCells.length;
    roomStartCells.push({
      ...cell,
      roomId: roomId,
    });
  }

  return roomStartCells;
};

