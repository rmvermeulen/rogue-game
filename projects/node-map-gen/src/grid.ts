import * as sort from 'alphanum-sort';
import chalk from 'chalk';
import * as os from 'os';
import {
  addIndex,
  allPass,
  append,
  clone,
  defaultTo,
  equals,
  filter,
  flatten,
  identity,
  ifElse,
  is,
  join,
  last,
  lte,
  map,
  pipe,
  pluck,
  prepend,
  propEq,
  range,
  repeat,
  tap,
  times,
  trim,
  tryCatch,
  uniq,
  when,
  where,
  whereEq,
} from 'ramda';
import { random } from './random';
import { generate } from './utils';

const colors = (() => {
  const src = [
    chalk.red,
    chalk.redBright,
    chalk.green,
    chalk.greenBright,
    chalk.blue,
    chalk.blueBright,
    chalk.yellow,
    chalk.yellowBright,
    chalk.magenta,
    chalk.magentaBright,
    chalk.cyan,
    chalk.cyanBright,
  ];

  return random.shuffle(src);
})();

export interface IGridOptions extends Partial<Grid> {
  width: number;
  height: number;
  rooms: number;
  cells?: ICell[];
}

export interface ICell {
  id: number;
  x: number;
  y: number;
  room: number;
}

export interface IRoom {
  id: number;
  size: number;
  cells: ICell['id'][];
}

// represents a grid of cells
export class Grid {
  public readonly width: number;
  public readonly height: number;
  public readonly cells: ICell[];
  private readonly rooms: number;
  private constructor(options: IGridOptions) {
    const { width, height, rooms, cells } = options;
    const isClone = options instanceof Grid;
    console.assert(
      isClone === Boolean(cells),
      'Invalid Grid.CLONE: incomplete source',
    );

    this.width = width;
    this.height = height;
    this.rooms = rooms;
    this.cells = isClone ? clone(cells) : this.generateCells();
  }
  public static CREATE(options: IGridOptions) {
    return new Grid({ ...options });
  }

  public static CLONE(grid: Grid) {
    // tslint:disable-next-line: no-any
    return new Grid(grid as any);
  }
  public clone() {
    return Grid.CLONE(this);
  }

  public cellAt(x: number, y: number): undefined | ICell {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.find((whereEq as any)({ x, y }));
  }

  // find all cells belonging to the given row
  public row(y: number): ICell[] {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.filter((whereEq as any)({ y }));
  }

  // find all cells belonging to the given column
  public column(x: number): ICell[] {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.filter((whereEq as any)({ x }));
  }

  // list the unique room-ids in the grid
  public listRooms(): number[] {
    return pipe<ICell[], number[], number[], number[]>(
      pluck('room'),
      uniq,
      filter(lte(0)),
    )(this.cells);
  }

  // get information on a specific room
  public findRoomById(id: IRoom['id']): IRoom | undefined {
    const cells = this.cells.filter(propEq('room', id));

    return cells.length > 0
      ? {
          id,
          size: cells.length,
          cells: pluck('id', cells),
        }
      : undefined;
  }

  /** render the grid into a string
   * example output:
   * +----+----+---+----+-----+
   * | 10   10 | 2   2  | 200 |
   * +----+    +   +    +-----+
   * | 0  | 10 | 2   2  | 10  |
   * +    +    +   +    +-----+
   * | 0  | 10 | 2   2  | 0   |
   * +    +    +   +    +     +
   * | 0  | 10 | 2   2  | 0   |
   * +    +----+---+----+-----+
   * | 0    0    0 | 10   10  |
   * +--- +----+---+----+-----+
   *
   * room 0 size=9 cells=4,5,10,14,15,19,20,21,22
   * room 1 size=8 cells=0,1,6,9,11,16,23,24
   * room 2 size=8 cells=2,3,7,8,12,13,17,18
   */
  // tslint:disable-next-line: max-func-body-length
  public display(useColors: boolean = true): string {
    const header = repeat('-', this.width);
    const lineFragments: (string | string[] | string[][])[] = [header];
    const columnNumberWidth: number[] = generate(this.width, () => 1);

    for (let y = 0; y < this.height; y += 1) {
      const line = [];
      const separatorLine = [];
      for (let x = 0; x < this.width; x += 1) {
        const cell = this.cellAt(x, y);
        const east = this.cellAt(x + 1, y);
        const south = this.cellAt(x, y + 1);

        // update the current column-width to the biggest in the column
        const roomStr = String(cell.room);
        columnNumberWidth[x] = Math.max(columnNumberWidth[x], roomStr.length);

        let str = roomStr;
        if (useColors) {
          const colorize = colors[cell.room % colors.length];
          str = colorize(roomStr);
        }

        const isSameRoom = when(Boolean, propEq('room', cell.room));
        line.push([
          str,
          (roomStr.length as unknown) as string /*cheat*/,
          isSameRoom(east) ? ' ' : '|',
        ]);
        separatorLine.push(isSameRoom(south) ? ' ' : '-');
      }
      lineFragments.push(line, separatorLine);
    }

    const rooms = sort(this.listRooms())
      .map(id => this.findRoomById(id))
      .filter(Boolean);

    if (rooms.length > 0) {
      lineFragments.push(
        [], // empty line to separate
        ...rooms.map(
          ({ id, size, cells }) => `room ${id} size=${size} cells=${cells}`,
        ),
      );
    }

    const mapIndexed = addIndex(map);
    const defaultTo1 = defaultTo(1);

    let previousWasVerticalWall = true;

    return pipe<typeof lineFragments, string[], string>(
      map(
        pipe(
          ifElse(
            is(String),
            identity,
            pipe(
              ifElse(
                ([first]) => is(Array, first),
                pipe(
                  mapIndexed(
                    (
                      [roomIdStr, roomIdStrLength, maybeWall]: [
                        string,
                        number,
                        '|' | ' '
                      ],
                      column,
                    ) => {
                      const width = defaultTo1(columnNumberWidth[column]);
                      // cannot use String.padEnd because of color in text
                      const paddedId =
                        roomIdStr + ' '.repeat(width - roomIdStrLength);

                      return `${paddedId} ${maybeWall} `;
                    },
                  ),
                  (list: string[]) => prepend('| ', list),
                ),
                pipe(
                  mapIndexed((wallType: '-' | ' ', column) => {
                    const width = defaultTo1(columnNumberWidth[column]);

                    const isVerticalWall = wallType === '-';

                    const dot =
                      isVerticalWall || previousWasVerticalWall
                        ? '+'
                        : column === 0
                        ? '|'
                        : ' ';

                    previousWasVerticalWall =
                      column === this.width - 1 ? isVerticalWall : false;

                    return dot + wallType.repeat(width + 2);
                  }),
                  (lines: string[]) =>
                    lines.length === 0
                      ? lines
                      : // try to read the last character of the last fragment
                      tryCatch(
                          pipe<string[], string, string>(
                            last,
                            last,
                          ),
                          () => ' ',
                        )(lines) === '-'
                      ? append('+', lines)
                      : append('|', lines),
                ),
              ),
              join(''),
            ),
          ),
          trim,
        ),
      ),
      join(os.EOL),
    )(lineFragments);
  }

  private generateCells() {
    const { width, height, rooms } = this;
    const avgRoomSize = Math.max(0, Math.floor((width * height) / rooms) - 1);
    const roomIds = times(n => repeat(n, avgRoomSize), rooms);
    let remainder = width * height - rooms * avgRoomSize;
    while (remainder > 0) {
      const id = random.natural({ max: rooms - 1 });
      roomIds[id].push(id);
      remainder -= 1;
    }

    const pool: number[] = random.shuffle(flatten(roomIds));

    const popSpecific = (n: number): number | undefined => {
      const i = pool.findIndex(equals(n));
      if (i === -1) {
        return undefined;
      }
      const [popped] = pool.splice(i, 1);

      return popped;
    };

    return pipe(
      map((id: number) => ({
        id,
        x: id % width,
        y: Math.floor(id / width),
        room: -1,
      })),
      tap((cells: ICell[]) => {
        cells.forEach((cell: ICell) => {
          const trySame = random.weighted([true, false], [15, 1]);
          let room: number;
          if (trySame) {
            // has valid room, which still has entries
            const neighbors = cells
              .filter(
                where({
                  room: allPass([
                    lte(0),
                    (r: ICell['room']) => pool.includes(r),
                  ]),
                }),
              )
              // manhattan distance
              .filter(
                ({ x, y }: typeof cell) =>
                  Math.abs(cell.x - x) + Math.abs(cell.y - y) === 1,
              );
            for (const n of neighbors) {
              const popped = popSpecific(n.room);
              if (popped !== undefined) {
                room = popped;
                break;
              }
            }
          }
          if (room === undefined) {
            room = pool.pop();
          }
          // update cell in-place
          cell.room = room;
        });
      }),
    )(range(0, width * height));
  }
}
