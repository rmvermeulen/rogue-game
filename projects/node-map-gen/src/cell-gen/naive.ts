import { assert } from 'chai';
import {
  allPass,
  equals,
  flatten,
  lte,
  map,
  pipe,
  range,
  repeat,
  tap,
  where,
} from 'ramda';
import { ICell, IGridOptions } from '../grid';
import { random } from '../random';
import { generate, natural } from '../utils';

// tslint:disable-next-line: no-empty-interface
export interface INaiveOptions extends IGridOptions {}

/**
 * generate a grid of cells divided into rooms
 * using a naive algorithm.
 * rooms can be spread into separate chunks as small as 1x1
 */
export const genCellsNaive = ({ width, height, rooms }: INaiveOptions) => {
  assert.equal(width, natural(width), 'width must be an integer > 0');
  assert.equal(height, natural(height), 'height must be a positive integer');
  assert.equal(rooms, natural(rooms), 'rooms must be a positive integer');

  const avgRoomSize = Math.max(0, Math.floor((width * height) / rooms) - 1);
  const roomIds = generate(rooms, n => repeat(n, avgRoomSize));
  let remainder = width * height - rooms * avgRoomSize;
  while (remainder > 0) {
    const id = random.natural({ max: rooms - 1 });
    roomIds[id].push(id);
    remainder -= 1;
  }

  const pool: number[] = random.shuffle(flatten(roomIds));

  const popSpecific = (n: number): number | undefined => {
    const i = pool.findIndex(equals(n));
    if (i === -1) {
      return undefined;
    }
    const [popped] = pool.splice(i, 1);

    return popped;
  };

  return pipe(
    map((id: number) => ({
      id,
      x: id % width,
      y: Math.floor(id / width),
      room: -1,
    })),
    tap((cells: ICell[]) => {
      cells.forEach((cell: ICell) => {
        const trySame = random.weighted([true, false], [15, 1]);
        let room: number;
        if (trySame) {
          // has valid room, which still has entries
          const neighbors = cells
            .filter(
              where({
                room: allPass([lte(0), (r: ICell['room']) => pool.includes(r)]),
              }),
            )
            // manhattan distance
            .filter(
              ({ x, y }: typeof cell) =>
                Math.abs(cell.x - x) + Math.abs(cell.y - y) === 1,
            );
          for (const n of neighbors) {
            const popped = popSpecific(n.room);
            if (popped !== undefined) {
              room = popped;
              break;
            }
          }
        }
        if (room === undefined) {
          room = pool.pop();
        }
        // update cell in-place
        cell.room = room;
      });
    }),
  )(range(0, width * height));
};
