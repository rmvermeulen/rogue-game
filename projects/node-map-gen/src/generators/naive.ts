import { assert } from 'chai';
import { allPass, equals, flatten, flip, gte, repeat, where } from 'ramda';
import { contained } from 'ramda-adjunct';
import { IGridOptions } from '../grid';
import { random } from '../random';
import { generate, natural } from '../utils';
import { manhattan } from './utils';

// tslint:disable-next-line: no-empty-interface
export interface INaiveOptions extends IGridOptions {}

/**
 * generate a grid of cells divided into rooms
 * using a naive algorithm.
 * rooms can be spread into separate chunks as small as 1x1
 */
export const generateCells = ({ width, height, roomCount }: INaiveOptions) => {
  assert.equal(width, natural(width), 'width must be an integer > 0');
  assert.equal(height, natural(height), 'height must be a positive integer');
  assert.equal(
    roomCount,
    natural(roomCount),
    'rooms must be a positive integer',
  );
  assert.isAtMost(roomCount, width * height, 'Too many rooms');

  const avgRoomSize = Math.max(0, Math.floor((width * height) / roomCount) - 1);
  const roomIds = generate(roomCount, n => repeat(n, avgRoomSize));
  let remainder = width * height - roomCount * avgRoomSize;
  while (remainder > 0) {
    const id = random.natural({ max: roomCount - 1 });
    roomIds[id].push(id);
    remainder -= 1;
  }

  const roomIdPool: number[] = random.shuffle(flatten(roomIds));

  const popSpecific = (n: number): number | undefined => {
    const i = roomIdPool.findIndex(equals(n));
    if (i === -1) {
      return undefined;
    }
    const [popped] = roomIdPool.splice(i, 1);

    return popped;
  };

  const cells = generate(width * height, (id: number) => ({
    id,
    x: id % width,
    y: Math.floor(id / width),
    roomId: -1,
  }));

  for (const cell of cells) {
    type CellData = typeof cell;
    // cells with valid room property
    // which is still included in pool
    const roomId = cells
      .filter(
        where({
          room: allPass([flip(gte)(0), contained(roomIdPool)]),
        }),
      )
      // manhattan distance
      .filter((other: CellData) => manhattan(cell, other) === 1)
      .reduce(
        (id: number | undefined, neighbor: CellData) =>
          id === undefined ? popSpecific(neighbor.roomId) : id,
        undefined,
      );

    // update cell in-place
    cell.roomId = roomId !== undefined ? roomId : roomIdPool.pop();
  }

  return cells;
};
