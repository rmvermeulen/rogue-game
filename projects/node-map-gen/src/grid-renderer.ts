import { assert } from 'chai';
import * as os from 'os';
import {
  aperture,
  compose,
  defaultTo,
  length,
  lensPath,
  map,
  path,
  pluck,
  repeat,
  set,
  sort,
  subtract,
  test,
  toString,
  unnest,
  when,
  zip,
} from 'ramda';
import {
  manyColors,
  manyColorsLength,
  someColors,
  someColorsLength,
} from './color-lists';
import { Grid, ICell } from './grid';
import { generate } from './utils';

export const renderSimple = (grid: Grid, useColors: boolean = true): string => {
  if (grid.listRooms().length > manyColorsLength) {
    // tslint:disable-next-line: no-parameter-reassignment
    useColors = false;
  }
  const colors = useColors
    ? grid.cells.length <= someColorsLength
      ? someColors()
      : manyColors()
    : [];

  const columnWidths = repeat(1, grid.width);

  generate(grid.height, (rowIndex: number) => {
    const row = grid.row(rowIndex);
    const roomStrWidths = compose<ICell[], number[], number[]>(
      map(
        compose<number, string, number>(
          length,
          toString,
        ),
      ),
      pluck<ICell, 'roomId'>('roomId'),
    )(row);
    // update column
    roomStrWidths.forEach((width, column) => {
      columnWidths[column] = Math.max(width, columnWidths[column]);
    });
  });

  return generate<string>(grid.height, (rowIndex: number) =>
    grid
      .row(rowIndex)
      .map((cell: ICell, index) => {
        let roomIdStr = cell.roomId.toString();
        const roomIdStrCharLength = roomIdStr.length;
        if (useColors) {
          const colorize = colors[cell.roomId % colors.length];
          roomIdStr = colorize(roomIdStr);
        }
        const width = columnWidths[index];

        // cannot use String.padEnd because of color in text
        return ' '.repeat(width - roomIdStrCharLength) + roomIdStr;
      })
      .join(' '),
  ).join(os.EOL);
};

// const stringifyLine = (
//   values: number[],
// ): { text: string; columnWidths: number[] } => {
//   return compose<number[], string[], { text: string; columnWidths: number[] }>(
//     applySpec({
//       text: (ss: string[]) => `| ${ss.join(' | ')} |`,
//       columnWidths: pluck('length'),
//     }),
//     map(String),
//   )(values);
// };

/** render the grid into a string
 * example output:
 * +----+----+---+----+-----+
 * | 10   10 | 2   2  | 200 |
 * +----+    +   +    +-----+
 * | 0  | 10 | 2   2  | 10  |
 * +    +    +   +    +-----+
 * | 0  | 10 | 2   2  | 0   |
 * +    +    +   +    +     +
 * | 0  | 10 | 2   2  | 0   |
 * +    +----+---+----+-----+
 * | 0    0    0 | 10   10  |
 * +--- +----+---+----+-----+
 *
 * room 0 size=9 cells=4,5,10,14,15,19,20,21,22
 * room 1 size=8 cells=0,1,6,9,11,16,23,24
 * room 2 size=8 cells=2,3,7,8,12,13,17,18
 */
// tslint:disable-next-line: max-func-body-length
export const render = (grid: Grid, useColors: boolean = true): string => {
  type WallStr = '|' | '+' | '-' | ' ';

  const columnWidths: number[] = generate(grid.width, () => 1);

  const mapLinesA = generate(grid.height, y => {
    const row = grid.row(y);
    const roomIds = pluck('roomId', row);
    generate(grid.width, column => {
      const tokenWidth = String(roomIds[column]).length;
      columnWidths[column] = Math.max(columnWidths[column], tokenWidth);
    });
    const walls: WallStr[] = [];
    for (const [a, b] of aperture(2, roomIds)) {
      const nextWall: WallStr = a === b ? ' ' : '|';
      walls.push(nextWall);
    }
    const [first, ...rest] = roomIds;

    return ['|', first, ...unnest(zip(walls, rest)), '|'];
  });

  const columnWalls: WallStr[][] = repeat([], grid.width);
  generate(grid.width, x => {
    const column = grid.column(x);
    const walls = columnWalls[x];
    walls.push('-');
    for (const [a, b] of aperture(2, pluck('roomId', column))) {
      const nextWall: WallStr = a === b ? ' ' : '-';
      walls.push(nextWall);
    }
    walls.push('-');
  });

  const topLine: WallStr[] = mapLinesA[0].map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const lastMapLine = mapLinesA[grid.height - 1];
  const bottomLine: WallStr[] = lastMapLine.map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const mapLinesB = [
    topLine,
    ...unnest(
      [...aperture(2, mapLinesA), [lastMapLine, undefined]].map(([a, b]) => {
        if (b === undefined) {
          return [a, bottomLine];
        }
        const walls = generate(a.length, x => {
          const ax = a[x];
          const bx = b[x];

          if (typeof ax === 'number') {
            return ax === bx ? ' ' : '-';
          }

          return ax === bx ? '|' : '+';
        });

        return [a, walls];
      }),
    ),
  ];

  assert.lengthOf(mapLinesB, grid.height * 2 + 1);

  let updated = mapLinesB;

  generate(grid.height + 1, gy => {
    generate(grid.width + 1, gx => {
      const x = gx * 2;
      const y = gy * 2;
      // const cornerPart = mapLinesB[y * 2][x * 2];
      const [n, s, e, w] = [[y - 1, x], [y + 1, x], [y, x - 1], [y, x + 1]]
        // tslint:disable-next-line: no-any no-unsafe-any
        .map(p => path(p as any[], mapLinesB))
        .map(defaultTo(' '));
      const assignField = (char: string) => {
        // tslint:disable-next-line: no-any
        const modify = set(lensPath([y, x]), char) as any;
        // tslint:disable-next-line: no-unsafe-any
        updated = modify(updated);
      };
      if (n === ' ' && s === ' ' && e === ' ' && w === ' ') {
        return assignField(' ');
      }
      if (e === ' ' && w === ' ') {
        return assignField('|');
      }
      if (n === ' ' && s === ' ') {
        return assignField('-');
      }

      return assignField('+');
    });
  });

  const lineFragments = updated.map((chunks, y) => {
    const newLine = chunks.map(String).map((chunk, x) => {
      const isRoomCenterX = x % 2 === 1;
      if (!isRoomCenterX) {
        return chunk;
      }
      const isRoomCenterY = y % 2 === 1;
      const width = defaultTo(1, columnWidths[(x - 1) / 2]);

      const padded = isRoomCenterY
        ? ` ${chunk.padStart(width)} `
        : chunk.repeat(width + 2);
      assert.lengthOf(padded, width + 2);

      return padded;
    });

    return newLine.join('');
  });

  const rooms = sort(subtract, grid.listRooms())
    .map(id => grid.findRoomById(id))
    .filter(Boolean);

  const roomFragments = rooms.map(
    ({ id, size, cells }) => `room ${id} size=${size} cells=${cells}`,
  );

  const renderedString = lineFragments.join(os.EOL);

  if (useColors) {
    const colors =
      grid.cells.length <= someColorsLength ? someColors() : manyColors();
    const getColorizer = (n: number) => colors[n % colors.length];

    return [
      renderedString
        .split(' ')
        .map(
          when(test(/^\d+$/), (str: string) => getColorizer(Number(str))(str)),
        )
        .join(' '),
      '',
      ...roomFragments,
    ].join(os.EOL);
  }

  return [renderedString, '', ...roomFragments].join(os.EOL);
};
