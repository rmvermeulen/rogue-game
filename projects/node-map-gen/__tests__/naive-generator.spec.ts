jest.doMock('../src/random');

import { contains, filter, pipe, pluck } from 'ramda';
import { generateCells, INaiveOptions } from '../src/generators/naive';
import { Grid, ICell } from '../src/grid';
import { render } from '../src/grid-renderer';
import { random } from '../src/random';
import { generate } from '../src/utils';

describe.each`
  width | height | roomCount
  ${5}  | ${5}   | ${4}
  ${5}  | ${8}   | ${8}
  ${8}  | ${5}   | ${12}
  ${8}  | ${8}   | ${20}
`(
  'naively generated grid (w $width h $height r $roomCount)',
  ({ width, height, roomCount }: INaiveOptions) => {
    let cells: ICell[];
    beforeAll(() => {
      // change the rng
      generate(1, () => random.natural());

      cells = generateCells({
        width,
        height,
        roomCount,
      });
    });

    it('is probably a grid', () => {
      expect(cells).toBeDefined();
      expect(cells).toHaveLength(width * height);
    });

    it.skip('does not have too many isolated cells', () => {
      const isolatedCells = cells.reduce((total, cell) => {
        const isPartOfRoom = pipe(
          filter(
            ({ x, y }: ICell) =>
              Math.abs(cell.x - x) + Math.abs(cell.y - y) === 1,
          ),
          pluck('roomId'),
          contains(cell.roomId),
        )(cells);

        return total + (isPartOfRoom ? 0 : 1);
      }, 0);
      expect(isolatedCells / cells.length).toBeLessThan(0.25);
    });

    it('looks fine', () => {
      const grid = Object.assign(Object.create(Grid.prototype) as Grid, {
        width,
        height,
        cells,
      });

      expect(render(grid, false)).toMatchSnapshot();
    });
  },
);
