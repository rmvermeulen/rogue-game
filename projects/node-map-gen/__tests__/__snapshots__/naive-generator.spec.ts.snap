// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`naively generated grid (w 5 h 5 r 4) looks fine 1`] = `
"+---+---+---+---+---+
| 0   0   0 | 3 | 0 |
+---        +---+---+
| 2 | 0   0 | 1   1 |
+---    +---+---+---+
| 3 | 0 | 2 | 3 | 2
+---+---+---+---+---+
| 2 | 3 | 1 | 2 | 3 |
+---+---    +---+---+
| 1   1   1 | 3 | 0 |
+---+---+---+---+---+

room 0 size=8 cells=0,1,2,4,6,7,11,24
room 1 size=6 cells=8,9,17,20,21,22
room 2 size=5 cells=5,12,14,15,18
room 3 size=6 cells=3,10,13,16,19,23"
`;

exports[`naively generated grid (w 5 h 8 r 8) looks fine 1`] = `
"+---+---+---+---+---+
| 2 | 0   0 | 4 | 6
+---+---+---+---+---+
| 6 | 1 | 6 | 5   5 |
+---    +---+---+---+
| 2 | 1 | 7 | 4 | 7
+---+---    +---+---+
| 7 | 6 | 7 | 3 | 0 |
+---+---+---+---+---+
| 1 | 0 | 3 | 6 | 4
+---+---+---+---    |
| 4 | 1 | 7 | 5 | 4 |
+---+---+---+---+---+
| 0 | 3 | 5 | 2   2 |
+---+---+---+---+---+
| 3 | 7 | 6   6 | 5 |
+---+---+---+---+---+

room 0 size=5 cells=1,2,19,21,30
room 1 size=4 cells=6,11,20,26
room 2 size=4 cells=0,10,33,34
room 3 size=4 cells=18,22,31,35
room 4 size=5 cells=3,13,24,25,29
room 5 size=5 cells=8,9,28,32,39
room 6 size=7 cells=4,5,7,16,23,37,38
room 7 size=6 cells=12,14,15,17,27,36"
`;

exports[`naively generated grid (w 8 h 5 r 12) looks fine 1`] = `
"+----+----+---+----+---+---+---+---+
| 6  | 3  | 6 | 10 | 4 | 5 | 4 | 5 |
+----+----+---     +---+---+---+---+
| 0  | 9  | 1 | 10 | 2 | 4 | 7 | 1 |
+----+----+---+----+---+---+---+---+
| 11   11 | 5 | 3  | 6   6 | 8   8
+----+----+---+----+---+---+---+---+
| 8  | 3  | 1 | 9  | 5 | 2 | 9 | 4 |
+----+----+---+----    +---+---+---+
| 7  | 8  | 2 | 0  | 5 | 9 | 2 | 7 |
+----+----+---+----+---+---+---+---+

room 0 size=2 cells=8,35
room 1 size=3 cells=10,15,26
room 2 size=4 cells=12,29,34,38
room 3 size=3 cells=1,19,25
room 4 size=4 cells=4,6,13,31
room 5 size=5 cells=5,7,18,28,36
room 6 size=4 cells=0,2,20,21
room 7 size=3 cells=14,32,39
room 8 size=4 cells=22,23,24,33
room 9 size=4 cells=9,27,30,37
room 10 size=2 cells=3,11
room 11 size=2 cells=16,17"
`;

exports[`naively generated grid (w 8 h 8 r 20) looks fine 1`] = `
"+----+----+----+----+----+----+----+----+
| 15 | 18 | 11 | 12 | 16 | 9  | 17 | 13 |
+----+----+----+----+----+----+----+----+
| 6  | 5  | 0  | 10 | 7  | 4  | 14 | 7  |
+----+----+----+----+----+----+----+----+
| 18 | 14 | 8  | 17 | 3  | 5  | 8  | 3  |
+    +----+----+----+----+----+----+----+
| 18 | 13 | 1  | 0  | 19 | 10 | 15 | 9  |
+    +----+----+----+----+----     +----+
| 18 | 4  | 15   15 | 6  | 16 | 15 | 12 |
+----+----+----+----+----+----+----+----+
| 2  | 15 | 9  | 19 | 2  | 12 | 4  | 18 |
+----+----+----+----+----+----+----+----+
| 17 | 10 | 14 | 6  | 17 | 4  | 1  | 9  |
+----+----+----+----+----+----+----+----+
| 1  | 7  | 13 | 14 | 8  | 6  | 8  | 11 |
+----+----+----+----+----+----+----+----+

room 0 size=2 cells=10,27
room 1 size=3 cells=26,54,56
room 2 size=2 cells=40,44
room 3 size=2 cells=20,23
room 4 size=4 cells=13,33,46,53
room 5 size=2 cells=9,21
room 6 size=4 cells=8,36,51,61
room 7 size=3 cells=12,15,57
room 8 size=4 cells=18,22,60,62
room 9 size=4 cells=5,31,42,55
room 10 size=3 cells=11,29,49
room 11 size=2 cells=2,63
room 12 size=3 cells=3,39,45
room 13 size=3 cells=7,25,58
room 14 size=4 cells=14,17,50,59
room 15 size=6 cells=0,30,34,35,38,41
room 16 size=2 cells=4,37
room 17 size=4 cells=6,19,48,52
room 18 size=5 cells=1,16,24,32,47
room 19 size=2 cells=28,43"
`;
